reviews/models/mixins.py

```diff
@@ -0,0 +1,193 @@
+        abstract = True
+
+    # NOTE: reviews_state should rarely/never be modified directly -- use the state transition methods below
+    reviews_state = models.CharField(max_length=15, db_index=True, choices=workflow.States.choices(), default=workflow.States.INITIAL.value)
+
+    date_last_transitioned = models.DateTimeField(null=True, blank=True, db_index=True)
+
+    @property
+    def in_public_reviews_state(self):
+        public_states = workflow.PUBLIC_STATES.get(self.provider.reviews_workflow)
+        if not public_states:
+            return False
+        return self.reviews_state in public_states
+
+    @property
+    def _reviews_machine(self):
+        if self.__machine is None:
+            self.__machine = ReviewsMachine(self, 'reviews_state')
+        return self.__machine
+
+    def reviews_submit(self, user):
+        """Run the 'submit' state transition and create a corresponding Action.
+
+        Params:
+            user: The user triggering this transition.
+        """
+        return self.__run_transition(self._reviews_machine.submit, user=user)
+
+    def reviews_accept(self, user, comment):
+        """Run the 'accept' state transition and create a corresponding Action.
+
+        Params:
+            user: The user triggering this transition.
+            comment: Text describing why.
+        """
+        return self.__run_transition(self._reviews_machine.accept, user=user, comment=comment)
+
+    def reviews_reject(self, user, comment):
+        """Run the 'reject' state transition and create a corresponding Action.
+
+        Params:
+            user: The user triggering this transition.
+            comment: Text describing why.
+        """
+        return self.__run_transition(self._reviews_machine.reject, user=user, comment=comment)
+
+    def reviews_edit_comment(self, user, comment):
+        """Run the 'edit_comment' state transition and create a corresponding Action.
+
+        Params:
+            user: The user triggering this transition.
+            comment: New comment text.
+        """
+        return self.__run_transition(self._reviews_machine.edit_comment, user=user, comment=comment)
+
+    def __run_transition(self, trigger_fn, **kwargs):
+        with transaction.atomic():
+            result = trigger_fn(**kwargs)
+            action = self._reviews_machine.action
+            if not result or action is None:
+                raise InvalidTransitionError()
+            return action
+
+
+class ReviewsMachine(Machine):
+
+    action = None
+    from_state = None
+
+    def __init__(self, reviewable, state_attr):
+        self.reviewable = reviewable
+        self.__state_attr = state_attr
+
+        super(ReviewsMachine, self).__init__(
+            states=[s.value for s in workflow.States],
+            transitions=workflow.TRANSITIONS,
+            initial=self.state,
+            send_event=True,
+            prepare_event=['initialize_machine'],
+            ignore_invalid_triggers=True,
```

Comment:
Some triggers are invalid depending on the current state -- you can't `submit` an already `accepted` preprint, or `reject` one that's still in `initial`. With this set `True`, the trigger method will return falsy instead of raising an error in those cases, so we can handle it with our own exception in `ReviewableMixin.__run_transition`.